File: 2264. Largest 3-Same-Digit Number in String/LargestGoodInteger.py
class Solution:
    
    # O(n) time,
    # O(1) space,
    # Approach: sliding window, hash map
    def largestGoodInteger(self, num: str) -> str:
        window = {}
        left, right = 0, 2
        ans = ""
        for i in range(right):
            digit = int(num[i])
            window[digit] = window.get(digit, 0) + 1
        
        while right < len(num):
            right_digit = int(num[right])
            window[right_digit] = window.get(right_digit, 0) + 1
            right += 1
            
            if len(window) == 1:
                if not ans or int(ans) < int(num[left:right]):
                    ans = num[left:right]
            
            left_digit = int(num[left])
            window[left_digit] -= 1
            if window[left_digit] == 0:
                window.pop(left_digit)
            left += 1
            
        return ans

File: 2264. Largest 3-Same-Digit Number in String/largest_good_integer.rs
impl Solution {

    // O(n) time,
    // O(n) space,
    // Approach: simulation, 
    pub fn largest_good_integer(num: String) -> String {
        let mut ans = String::from("");
        let mut max_num = -1;
        let digits: Vec<char> = num.chars().collect();

        let mut index = 0 as usize;
        while index < digits.len()-2 {
            if digits[index] == digits[index+1] && digits[index] == digits[index+2] {
                let substring: String = digits[index..index+3].iter().collect();
                let num: i32 = substring.parse().unwrap();

                if num > max_num {
                    max_num = num;
                    ans = substring;
                }
                index += 2;
            }
            index += 1;
        }

        ans
    }   
}